processing/src/main/java/org/apache/druid/frame/processor/SuperSorter.java

```diff
@@ -0,0 +1,867 @@
+  private final Queue<Frame> inputBuffer = new ArrayDeque<>();
+
+  @GuardedBy("runWorkersLock")
+  private long inputFramesReadSoFar = 0;
+
+  @GuardedBy("runWorkersLock")
+  private long levelZeroMergersRunSoFar = 0;
+
+  @GuardedBy("runWorkersLock")
+  private int ultimateMergersRunSoFar = 0;
+
+  @GuardedBy("runWorkersLock")
+  private final Map<File, FrameFile> penultimateFrameFileCache = new HashMap<>();
+
+  @GuardedBy("runWorkersLock")
+  private SettableFuture<OutputChannels> allDone = null;
+
+  @GuardedBy("runWorkersLock")
+  SuperSorterProgressTracker superSorterProgressTracker;
+
+  /**
+   * See {@link #setNoWorkRunnable}.
+   */
+  @GuardedBy("runWorkersLock")
+  private Runnable noWorkRunnable = null;
+
+  /**
+   * Initializes a SuperSorter.
+   *
+   * @param inputChannels              input channels. All frames in these channels must be sorted according to the
+   *                                   {@link ClusterBy#getColumns()}, or else sorting will not produce correct
+   *                                   output.
+   * @param frameReader                frame reader for the input channels
+   * @param clusterBy                  desired sorting order
+   * @param outputPartitionsFuture     a future that resolves to the desired output partitions. Sorting will block
+   *                                   prior to writing out final outputs until this future resolves. However, the
+   *                                   sorter will be able to read all inputs even if this future is unresolved.
+   *                                   If output need not be partitioned, use
+   *                                   {@link ClusterByPartitions#oneUniversalPartition()}. In this case a single
+   *                                   sorted channel is generated.
+   * @param exec                       executor to perform work in
+   * @param temporaryDirectory         directory to use for scratch files. This must have enough space to store at
+   *                                   least two copies of the dataset in {@link FrameFile} format.
+   * @param outputChannelFactory       factory for partitioned, sorted output channels
+   * @param innerFrameAllocatorMaker   supplier for allocators that are used to make merged frames for intermediate
+   *                                   levels of merging, prior to the final output. Final output frame allocation is
+   *                                   controlled by outputChannelFactory. One allocator is created per intermediate
+   *                                   scratch file.
+   * @param maxActiveProcessors        maximum number of merging processors to execute at once in the provided
+   *                                   {@link FrameProcessorExecutor}
+   * @param maxChannelsPerProcessor    maximum number of channels to merge at once per merging processor
+   * @param rowLimit                   limit to apply during sorting. The limit is merely advisory: the actual number
```

Comment:
Is this a soft limit on the final number of rows that _get output post sorting_? Or is this a limit on the number of rows per partition?

Reply:
It's applied across all partitions, not to each partition individually. I added that sentence to the javadoc to hopefully clarify it.